### PR TITLE
[CP] Fuse zigzag attention into a single call

### PR DESCRIPTION
**Pull request by Fridge003**
## Motivation

Split PR 3 of #32714. Fuse the two TRT-LLM attention calls in zigzag context parallelism into one independent optimization PR.

## Modifications

- build combined zigzag TRT-LLM attention metadata
- represent previous-rank and next-rank attention as one synthetic batch
- prepare duplicated TRT-LLM page tables once per forward, with a TODO to remove the duplicated storage
- invoke TRT-LLM context attention once and split the combined result
- leave the original non-TRT-LLM zigzag attention path unchanged
- keep one focused unit test comparing the combined result with the two-half reference

This PR was split from #32714 and then updated according to review feedback. It now changes 3 files.

## Accuracy Tests

- `py_compile` passed for all 3 changed Python files.
- All pre-commit hooks passed.
- Focused pytest collection was blocked on this macOS checkout by missing SGLang runtime dependencies (the current missing module is `PIL`); functional tests are not reported as passing locally.

## Speed Tests and Profiling

The source profile reduced the per-rank TRT-LLM context-attention kernel count from 72 to 36 (288 to 144 across four ranks). No benchmark was rerun for this review update; see #32714 for the original profile and end-to-end measurements.

## Checklist

- [x] Format code with pre-commit.
- [x] Keep only the requested focused unit test.
- [x] Preserve the original non-TRT-LLM attention path.
- [x] Follow the SGLang code style guidance.

---
### CI States

Latest PR Test (Base): :x: [Run #30780923583](https://github.com/sgl-project/sglang/actions/runs/30780923583)
Latest PR Test (Extra): :x: [Run #30780923408](https://github.com/sgl-project/sglang/actions/runs/30780923408)

### CODE REVIEW

File: test/registered/unit/layers/attention/test_trtllm_mha_zigzag.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
+import unittest
```
Comment (Fridge003) [inline test/registered/unit/layers/attention/test_trtllm_mha_zigzag.py]: Remove this test

File: python/sglang/srt/layers/cp/zigzag.py [hunk 2]
```diff
@@ -327,31 +353,33 @@ def run_attention(
         ), f"{self.name} CP does not support {attention_backend=}"
 
         meta = forward_batch.attn_cp_metadata
-        q_prev = q[: meta.total_q_prev_tokens]
         logical_tokens = meta.total_q_prev_tokens + meta.total_q_next_tokens
-        q_next = q[meta.total_q_prev_tokens : logical_tokens]
-
-        prev_kwargs = {}
```
Comment (Fridge003) [inline python/sglang/srt/layers/cp/zigzag.py:334]: prev_kwargs and next_kwargs shouldn't be removed when attn backend is not trtllm_mha
Don't modify the logic of original path

File: test/registered/cp/test_cp_strategy_unit.py [hunk 3]
```diff
@@ -547,6 +624,197 @@ def attn_fn(q_chunk, cu_seqlens_q, cache_seqlens, max_seqlen_q):
...
+            attn_fn=attn_fn,
+            attention_backend=CPAttentionBackendKind.TRTLLM_MHA,
+        )
+
+        self.assertEqual(len(calls), 1)
+        self.assertTrue(torch.equal(calls[0], logical_q))
+        self.assertTrue(torch.equal(out[:4], logical_q + 100))
+        self.assertTrue(torch.equal(out[4:], torch.zeros_like(out[4:])))
+
+    def test_zigzag_combined_attention_matches_two_half_reference(self):
```
Comment (Fridge003) [inline test/registered/cp/test_cp_strategy_unit.py:550]: We only need this unit test. Please delete any other unit tests added in this PR

File: python/sglang/srt/layers/attention/trtllm_mha_backend.py [hunk 4]
```diff
@@ -339,6 +342,35 @@ def _get_layer_page_table(
                 return swa_pt
         return self.forward_metadata.page_table
 
+    def _build_zigzag_page_tables(
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/trtllm_mha_backend.py]: Rename this function to `_maybe_build_cp_zigzag_page_tables`

File: python/sglang/srt/layers/attention/trtllm_mha_backend.py [hunk 5]
```diff
@@ -339,6 +342,35 @@ def _get_layer_page_table(
...
+
+        metadata.zigzag_page_table = torch.cat(
+            (metadata.page_table, metadata.page_table), dim=0
+        )
+        if metadata.swa_page_table is not None:
+            metadata.zigzag_swa_page_table = torch.cat(
+                (metadata.swa_page_table, metadata.swa_page_table), dim=0
+            )
+
+    def _get_zigzag_layer_page_table(
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/trtllm_mha_backend.py]: No need to create this function. The logics can be put under _trtllm_context_attn inline

File: python/sglang/srt/layers/attention/trtllm_mha_backend.py [hunk 6]
```diff
@@ -339,6 +342,35 @@ def _get_layer_page_table(
                 return swa_pt
         return self.forward_metadata.page_table
 
+    def _build_zigzag_page_tables(
+        self,
+        metadata: TRTLLMMHAMetadata,
+        forward_batch: ForwardBatch,
+    ) -> None:
+        """Duplicate request rows once for the combined prev-then-next CP launch."""
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/trtllm_mha_backend.py:350]: Add a TODO on we plan to save the memory usage for zigzag page table

**Comment by Fridge003** [comment]
/rerun-test test/registered/cp/test_cp_strategy_unit.py test/registered/cp/test_gqa_prefill_cp.py test/registered/cp/test_mimo_cp.py test/registered/cp/test_gpt_oss_4gpu_mxfp4_cp.py
